core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java

```diff
@@ -363,29 +369,102 @@ public long requestTimeoutMillis() {
+        if (requestTimeoutMillis == 0) {
+            clearRequestTimeout();
+        }
+
+        final long adjustmentMillis =
+                LongMath.saturatedSubtract(requestTimeoutMillis, this.requestTimeoutMillis);
+        extendRequestTimeoutMillis(adjustmentMillis);
+    }
+
     @Override
     public void setRequestTimeout(Duration requestTimeout) {
         setRequestTimeoutMillis(requireNonNull(requestTimeout, "requestTimeout").toMillis());
     }
 
+    @Override
+    public void extendRequestTimeoutMillis(long adjustmentMillis) {
+        if (adjustmentMillis == 0 || requestTimeoutMillis == 0) {
+            return;
+        }
+
+        final long oldRequestTimeoutMillis = requestTimeoutMillis;
+        requestTimeoutMillis = LongMath.saturatedAdd(oldRequestTimeoutMillis, adjustmentMillis);
+        final TimeoutController requestTimeoutController = this.requestTimeoutController;
+        if (requestTimeoutController != null) {
+            if (ch.eventLoop().inEventLoop()) {
+                requestTimeoutController.extendTimeout(adjustmentMillis);
+            } else {
+                ch.eventLoop().execute(() -> requestTimeoutController.extendTimeout(adjustmentMillis));
+            }
+        }
+    }
+
+    @Override
+    public void extendRequestTimeout(Duration adjustment) {
+        extendRequestTimeoutMillis(requireNonNull(adjustment, "adjustment").toMillis());
+    }
+
+    @Override
+    public void setRequestTimeoutAfterMillis(long requestTimeoutMillis) {
+        checkArgument(requestTimeoutMillis > 0,
+                      "requestTimeoutMillis: " + requestTimeoutMillis + " (expected: > 0)");
+
+        long passedTimeMillis = 0;
+        final TimeoutController requestTimeoutController = this.requestTimeoutController;
+        if (requestTimeoutController != null) {
+            passedTimeMillis = TimeUnit.NANOSECONDS.toMillis(
+                    System.nanoTime() - requestTimeoutController.startTimeNanos());
+            if (ch.eventLoop().inEventLoop()) {
+                requestTimeoutController.resetTimeout(requestTimeoutMillis);
+            } else {
+                ch.eventLoop().execute(() -> requestTimeoutController
+                        .resetTimeout(requestTimeoutMillis));
+            }
+        }
+
+        this.requestTimeoutMillis = LongMath.saturatedAdd(passedTimeMillis, requestTimeoutMillis);
+    }
+
+    @Override
+    public void setRequestTimeoutAfter(Duration requestTimeout) {
+        setRequestTimeoutAfterMillis(requireNonNull(requestTimeout, "requestTimeout").toMillis());
+    }
+
+    @Override
+    public void setRequestTimeoutAtMillis(long requestTimeoutAtMillis) {
+        checkArgument(requestTimeoutAtMillis >= 0,
+                      "requestTimeoutAtMillis: " + requestTimeoutAtMillis + " (expected: >= 0)");
+        final long requestTimeoutAfter = requestTimeoutAtMillis - Instant.now().toEpochMilli();
```

Comment:
Shouldn't we check the `requestTimeoutAfter` here again so that a user get meaningful message instead of `"requestTimeoutMillis: -12345 (expected: > 0)")`